agent/consul/state/catalog.go

```diff
@@ -4628,6 +4690,67 @@ func (s *Store) combinedServiceNodesTxn(tx ReadTxn, ws memdb.WatchSet, names []s
 	return maxIdx, resp, nil
 }
 
+func upstreamsFromGWTxn(tx ReadTxn, service structs.ServiceName) ([]structs.ServiceName, error) {
+	val, err := tx.First(tableConfigEntries, indexID, configentry.KindName{Kind: structs.BoundAPIGateway, Name: service.Name})
+	if err != nil {
+		return nil, err
+	}
+
+	if gw, ok := val.(*structs.BoundAPIGatewayConfigEntry); ok {
+		serviceIDs := gw.ListRelatedServices()
+		names := make([]structs.ServiceName, 0, len(serviceIDs))
+		for _, id := range serviceIDs {
+			names = append(names, structs.NewServiceName(id.ID, &id.EnterpriseMeta))
```

Comment:
this assumes that `id.ID` is the service name, since we set that directly when we create the `Services` field on the BoundAPIGateway I think it's okay to assume that